proc-macros/src/new/lifetimes.rs

```diff
@@ -0,0 +1,118 @@
+use proc_macro2::Span;
+use syn::spanned::Spanned;
+use syn::{Lifetime, Path, Type};
+
+static LIFETIME: &str = "'a";
+
+pub fn replace_lifetimes(ty: &mut Type) -> bool {
+	let mut replaced = false;
+	traverse_type(ty, &mut replaced, replace_lifetime);
+	replaced
+}
+
+pub fn replace_lifetime(ty: &mut Type, replaced: &mut bool) {
+	use syn::{GenericArgument, PathArguments};
+
+	match ty {
+		Type::Path(p) => {
+			for segment in p.path.segments.iter_mut() {
+				if let PathArguments::AngleBracketed(ref mut ab) = segment.arguments {
+					let mut lifetimes = 0;
+
+					for arg in ab.args.iter_mut() {
+						match arg {
+							GenericArgument::Lifetime(lt) => {
+								*lt = Lifetime::new(LIFETIME, lt.span());
+								*replaced = true;
+								lifetimes += 1;
+							}
+							// Stop iterating on first non-lifetime generic argument
+							// TODO: Replace by `.iter_mut().map_while(...).count()` when it's stabilized
+							_ => break,
+						}
+					}
+
+					// Check if the type is a `Cow<str>` or similar with lifetime elision,
+					// if so insert the lifetime as first argument.
+					if lifetimes == 0 && segment.ident == "Cow" {
```

Comment:
Could `lifetimes` be a bool? If I'm understanding this right, we can only swap at most one lifetime param out, and after that we'd ideally throw some sort of error saying that multiple lifetimes aren't supported or something.